### PR TITLE
[WIP] Initial Rollup local dev

### LINKED ISSUE #1: Setup local dev
https://storybook.js.org/docs/vue/get-started/introduction

### PR DESCRIPTION
**Pull request by tuhuynh27**
Resolve #1 

- [Reference](https://dev.to/shubhadip/vue-3-component-library-270p)

Initial (similar structure as `moai/react`):

- Yarn workspace for code sharing
- Rollup build for `core`
- Vite + Storybook for `docs`

Remaining issues:

Components must be imported [like this](https://github.com/moaijs/moai-vue/blob/feat%2Frollup-local-dev/lib/docs/src/stories/components/Button.stories.js#L1):
```vue
import { Button } from '@moai-vue/core/dist/esm/button'
```
Maybe we should update the correct rollup build 🤔

### CODE REVIEW

File: lib/core/postcss.config.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      }
+    },
+    "postcss-simple-vars": {},
+    "postcss-nested": {},
+    "postcss-url": {},
+    autoprefixer: {
+      overrideBrowserslist: "> 1%, IE 6, Explorer >= 10, Safari >= 7"
+    }
+  }
+};
```
Comment (ng-hai) [inline lib/core/postcss.config.js]: We could use [PostCSS Preset Env ](https://github.com/csstools/postcss-preset-env) that already includes most of features like CSS Variables, Nesting, Autoprefixer, etc. out of the box.

Comment (tuhuynh27) [reply lib/core/postcss.config.js]: Already used it in rollup build, see [https://github.com/moaijs/moai-vue/blob/8427e2286f592244ff66364f3b65ce62c690b7e4/lib/core/rollup.config.js#L9](https://github.com/moaijs/moai-vue/blob/8427e2286f592244ff66364f3b65ce62c690b7e4/lib/core/rollup.config.js#L9)

File: lib/core/postcss.config.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    "postcss-import": {
+      resolve(id, basedir) {
+        // resolve alias @css, @import '@css/style.css'
+        // because @css/ has 5 chars
+        if (id.startsWith("@css")) {
+          return path.resolve("./src/assets/styles/css", id.slice(5));
+        }
+
+        // resolve node_modules, @import '~normalize.css/normalize.css'
+        // similar to how css-loader's handling of node_modules
```
Comment (thien-do) [inline lib/core/postcss.config.js]: yo I didn't even know CSS loader can handle node_modules with the "~" sign 👍

File: lib/core/postcss.config.js (hunk 1 above)
Comment (thien-do) [reply lib/core/postcss.config.js]: so are we good here @ng-hai ?

File: lib/core/rollup.config.js [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+  postcssImport({
+    resolve(id, basedir) {
+      // resolve alias @css, @import '@css/style.css'
+      // because @css/ has 5 chars
+      if (id.startsWith("@css")) {
+        return path.resolve("./src/assets/styles/css", id.slice(5));
+      }
+
+      // resolve node_modules, @import '~normalize.css/normalize.css'
+      // similar to how css-loader's handling of node_modules
```
Comment (thien-do) [inline lib/core/rollup.config.js]: Wait is it dejavu.. I think I read it somewhere above? Are they 2 different setups?

File: lib/core/rollup.config.js [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+
+      // resolve relative path, @import './components/style.css'
+      return path.resolve(basedir, id);
+    }
+  }),
+  simplevars,
+  nested,
+  postcssUrl({ url: "inline" }),
+  autoprefixer({
+    overrideBrowserslist: "> 1%, IE 6, Explorer >= 10, Safari >= 7"
```
Comment (thien-do) [inline lib/core/rollup.config.js]: Are you sure about this? This seems.. very old

File: lib/core/rollup.config.js [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+      if (id.startsWith("~")) {
+        return path.resolve("./node_modules", id.slice(1));
+      }
+
+      // resolve relative path, @import './components/style.css'
+      return path.resolve(basedir, id);
+    }
+  }),
+  simplevars,
+  nested,
```
Comment (thien-do) [inline lib/core/rollup.config.js]: To be honest I'm not a fan of using many plugins for postcss (this is not a change request)

File: lib/core/rollup.config.js [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+      generateScopedName: '[local]___[hash:base64:5]',
+    },
+    include: /&module=.*\.css$/,
+  }),
+  // Process all `<style>` blocks except `<style module>`.
+  PostCSS({ include: /(?<!&module=.*)\.css$/,
+    plugins:[
+      ...postcssConfigList
+    ]
+  }),
```
Comment (thien-do) [inline lib/core/rollup.config.js]: I got this but I always thought this is.. built-in

File: lib/core/rollup.config.js [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+  url({
+    include: [
+      '**/*.svg',
+      '**/*.png',
+      '**/*.gif',
+      '**/*.jpg',
+      '**/*.jpeg'
+    ]
+  }),
```
Comment (thien-do) [inline lib/core/rollup.config.js]: I don't think we should bring along bitmap images like PNG or JPG. Thing is we either:
- Leave them as is (binary files), which means the users may need to setup proper loaders, even for files inside node_modules. Some tools like Nextjs will complain about this
- Or we bundled them inside JS code, which means they will be inlined as base64 or something, which makes the sizing a little bit hard to control

Anyway, my point is, I don't know how to handle bitmap image files in a distributable libs. If you know what you're doing here, just go ahead

File: lib/core/src/components/button/Button.vue [hunk 8]
```diff
@@ -0,0 +1,18 @@
+<template>
+  <button>{{ label }}</button>
+</template>
+
+<script lang="ts">
+import { defineComponent, ref } from "vue";
+export default defineComponent({
```
Comment (thien-do) [inline lib/core/src/components/button/Button.vue:7]: Hm I'm not sure if we can use named export with Vue. Generally I'd prefer named exports to avoid implicit naming. (i.e. if the users want to use Button as some other names, they must explicitly rename their imports)

File: lib/core/src/components/button/index.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+  install(vue: Application) {
+    registerComponent(vue, Button);
+  }
+};
+
+// Insert Vue use(Plugin);
+
+export default Plugin;
+
+export { Button };
```
Comment (thien-do) [inline lib/core/src/components/button/index.ts:16]: Hm I'm not really understand this (due to my lack of Vue knowledge). Is there a reference I should read to understand this "Plugin"?

File: lib/core/src/components/index.ts [hunk 10]
```diff
@@ -0,0 +1,3 @@
+import Button from './button'
```
Comment (thien-do) [inline lib/core/src/components/index.ts]: So this "Button" is importing the default "Plugin" export, not the named "Button" export?

File: lib/core/tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,36 @@
+{
+  "compilerOptions": {
+    "target": "esnext",
```
Comment (thien-do) [inline lib/core/tsconfig.json]: Are you sure about this? It's weird to see our browserlist supports IE6 but the TS is configured to ESNext

File: lib/core/tsconfig.json [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    "module": "esnext",
+    "strict": true,
+    "jsx": "preserve",
```
Comment (thien-do) [inline lib/core/tsconfig.json]: oh Vue also use JSX?

File: lib/docs/src/stories/Introduction.stories.mdx [hunk 13]
```diff
@@ -0,0 +1,211 @@
+import { Meta } from '@storybook/addon-docs/blocks';
+import Code from './assets/code-brackets.svg';
+import Colors from './assets/colors.svg';
+import Comments from './assets/comments.svg';
+import Direction from './assets/direction.svg';
+import Flow from './assets/flow.svg';
+import Plugin from './assets/plugin.svg';
+import Repo from './assets/repo.svg';
+import StackAlt from './assets/stackalt.svg';
```
Comment (thien-do) [inline lib/docs/src/stories/Introduction.stories.mdx]: Hm this is a lot of things. Are you sure we want this default Introduction?

File: lib/core/src/components/button/index.ts (hunk 9 above)
Comment (tuhuynh27) [reply lib/core/src/components/button/index.ts:16]: @thien-do Please refer this https://v3.vuejs.org/guide/plugins.html

Comment (tuhuynh27) [reply lib/core/src/components/button/index.ts:16]: In Vue, we can just `install` components once, not need to `import` at every view page

**Review by thien-do — COMMENTED** [review]
Well I have 2 main comments:
1. I think we can remove the "Introduction" story to make this PR more reviewable (or you can even go further and extract the "docs" setup to another PR)
2. The Rollup config is quite huge at the moment. Can we have it incrementally? This is not a strict change request, just consider if we can incrementally improving our config instead of having them all here.

File: lib/core/src/components/button/Button.vue (hunk 8 above)
Comment (monodyle) [reply lib/core/src/components/button/Button.vue:7]: Maybe you want something like

```
const Button = defineComponent({
...

export { Button } 
```

right?

Comment (iamquang95) [reply lib/core/src/components/button/Button.vue:7]: Agree on this one, i prefer named export which is easier for vscode auto import

File: lib/core/tsconfig.json (hunk 12 above)
Comment (monodyle) [reply lib/core/tsconfig.json]: Yes, it is. Reference: https://v3.vuejs.org/guide/render-function.html#jsx

File: lib/core/src/components/button/Button.vue (hunk 8 above)
Comment (thien-do) [reply lib/core/src/components/button/Button.vue:7]: yub ^

Comment (0x2c7) [reply lib/core/src/components/button/Button.vue:7]: *yup

**Comment by tuhuynh27** [comment]
As mentioned in PR desc, the `rollup` configs are copied from [this article](https://dev.to/shubhadip/vue-3-component-library-270p)

We can simplify it due to our needs, by the way, I'm stucking at:
```
Error: PostCSS plugin postcss-import requires PostCSS 8.
```
When running `yarn build`, help 😢

**Comment by monodyle** [comment]
> As mentioned in PR desc, the `rollup` configs are copied from [this article](https://dev.to/shubhadip/vue-3-component-library-270p)
> 
> We can simplify it due to our needs, by the way, I'm stucking at:
> 
> ```
> Error: PostCSS plugin postcss-import requires PostCSS 8.
> ```
> 
> When running `yarn build`, help 😢

Deja vu, I've stuck in this error a long time ago. The resolved way is to downgrade to postcss 7 (In my timeline). But who knows what @thien-do wants to do? 🦘

**Comment by iamquang95** [comment]
> As mentioned in PR desc, the `rollup` configs are copied from [this article](https://dev.to/shubhadip/vue-3-component-library-270p)
> 
> We can simplify it due to our needs, by the way, I'm stucking at:
> 
> ```
> Error: PostCSS plugin postcss-import requires PostCSS 8.
> ```
> 
> When running `yarn build`, help 😢

I faced this before on my personal project. I had to downgrade to PostCss 7 (https://tailwindcss.com/docs/installation#post-css-7-compatibility-build)

File: lib/core/rollup.config.js [hunk 14]
```diff
@@ -0,0 +1,280 @@
+import fs from "fs";
+import path from "path";
+import vue from "rollup-plugin-vue";
+import alias from "@rollup/plugin-alias";
+import commonjs from "@rollup/plugin-commonjs";
+import resolve from "@rollup/plugin-node-resolve";
+import replace from "@rollup/plugin-replace";
+import babel from "@rollup/plugin-babel";
+import PostCSS from "rollup-plugin-postcss";
```
Comment (thien-do) [inline lib/core/rollup.config.js]: @tuhuynh27 @monodyle is it possible to NOT having PostCSS at first? I mean we will eventually need it, but setting it up is not simple due to the huge breaking change of PostCSS 8.

You can use the "postcss-combat" solution, but it would significantly increase the complexity of this PR.. so my suggestion is to just remove it for now

File: lib/core/rollup.config.js [hunk 15]
```diff
@@ -0,0 +1,280 @@
+import fs from "fs";
+import path from "path";
+import vue from "rollup-plugin-vue";
+import alias from "@rollup/plugin-alias";
+import commonjs from "@rollup/plugin-commonjs";
+import resolve from "@rollup/plugin-node-resolve";
+import replace from "@rollup/plugin-replace";
+import babel from "@rollup/plugin-babel";
```
Comment (thien-do) [inline lib/core/rollup.config.js]: hm are you sure we need babel? I mean using the "lib" and "target" options in TSconfig can help us pretty much

File: lib/core/rollup.config.js [hunk 16]
```diff
@@ -0,0 +1,280 @@
+import fs from "fs";
+import path from "path";
+import vue from "rollup-plugin-vue";
+import alias from "@rollup/plugin-alias";
+import commonjs from "@rollup/plugin-commonjs";
+import resolve from "@rollup/plugin-node-resolve";
```
Comment (thien-do) [inline lib/core/rollup.config.js]: I really think you should read this config and simplify it much more :D the "commonjs", for example, you should not need, because we won't bundle 3rd party libraries inside our lib. This plugin is only used for "app", not "lib"

File: lib/core/postcss.config.js (hunk 1 above)
Comment (ng-hai) [reply lib/core/postcss.config.js]: No, I mean we won't need to install these plugins if applied `postcss-preset-env`
- postcss-simple-vars
- postcss-nested 
- autoprefixer

File: lib/core/rollup.config.js (hunk 14 above)
Comment (tuhuynh27) [reply lib/core/rollup.config.js]: Okay

File: lib/core/rollup.config.js [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+  }
+};
+
+// ESM/UMD/IIFE shared settings: externals
+// Refer to https://rollupjs.org/guide/en/#warning-treating-module-as-external-dependency
+const external = [
+  // list external dependencies, exactly the way it is written in the import statement.
+  // eg. 'jquery'
+  "vue"
+];
```
Comment (thien-do) [inline lib/core/rollup.config.js]: This is good

File: lib/core/rollup.config.js [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+
+// UMD/IIFE shared settings: output.globals
+// Refer to https://rollupjs.org/guide/en#output-globals for details
+const globals = {
+  // Provide global variable names to replace your external imports
+  // eg. jquery: '$'
+  vue: "Vue"
+};
```
Comment (thien-do) [inline lib/core/rollup.config.js]: FYI Moai doesn't ship with an UMD version =)) I just think it's too.. old

File: lib/core/src/components/button/Button.vue (hunk 8 above)
Comment (iamquang95) [reply lib/core/src/components/button/Button.vue:7]: Other option:
```
@Options({
	...
})
export default class Component extends Vue {}
```

I don't know which one is better.

File: lib/docs/storybook-static/11.7f35c06cd77f40a88403.manager.bundle.js.LICENSE.txt [hunk 19]
```diff
@@ -0,0 +1,12 @@
+/*!
```
Comment (thien-do) [inline lib/docs/storybook-static/11.7f35c06cd77f40a88403.manager.bundle.js.LICENSE.txt]: Pretty much better already. Just don't forget to gitignore the build of storybook

File: lib/docs/.gitignore [hunk 20]
```diff
@@ -0,0 +1,6 @@
+node_modules
+.DS_Store
+dist
+dist-ssr
+*.local
+storybook-static
```
Comment (huyng12) [inline lib/docs/.gitignore]: I think adding new line at the end of file is a good practice.

Comment (thien-do) [reply lib/docs/.gitignore]: ah yes, me too, I think we can address it using prettier or something similar in another PR?

Comment (tuhuynh27) [reply lib/docs/.gitignore]: Yeah, next PR will be the prettier + CI linter

File: lib/core/src/index.ts [hunk 21]
```diff
@@ -0,0 +1,16 @@
+import { App as Application, Plugin } from "vue";
+import * as components from "./components/index";
```
Comment (huyng12) [inline lib/core/src/index.ts:2]: I see you are using both relative path and absolute path. I suggest using only absolute path for consistent.

File: lib/core/postcss.config.js (hunk 1 above)
Comment (thien-do) [reply lib/core/postcss.config.js]: Yeah I think we will deal with CSS/Styling in another PR :)) just a JS-only build is good enough for the first PR

File: lib/core/rollup.config.js [hunk 22]
```diff
@@ -0,0 +1,146 @@
+import fs from "fs";
```
Comment (thien-do) [inline lib/core/rollup.config.js:1]: This config still has some unused parts I think, but it's much better and I think we can merge this after you ensure we can run a simple local dev.

Actually the "rollup" should be for distribution, so it's not really required to be completely correct atm

File: lib/core/src/components/button/Button.vue (hunk 8 above)
Comment (tuhuynh27) [reply lib/core/src/components/button/Button.vue:7]: Please refer here @thien-do [https://github.com/moaijs/moai-vue/blob/143d98db3a7d189e477fc36048592fc62f57feb2/lib/core/src/components/button/index.ts#L16](https://github.com/moaijs/moai-vue/blob/143d98db3a7d189e477fc36048592fc62f57feb2/lib/core/src/components/button/index.ts#L16)

Already required to use named import, to import a component, we'll need to import like this [https://github.com/moaijs/moai-vue/blob/143d98db3a/lib/docs/src/stories/components/Button.stories.js#L1](https://github.com/moaijs/moai-vue/blob/143d98db3a/lib/docs/src/stories/components/Button.stories.js#L1)

**Review by thien-do — APPROVED** [review]
It's pretty much better, and the PR is long enough. It's starting to be hard to keep track of all conversations. I think we can merge this now and continue other works (CI, linter, styling)

And also please try to simplify the setup even more :))

File: lib/core/src/components/button/Button.vue (hunk 8 above)
Comment (tuhuynh27) [reply lib/core/src/components/button/Button.vue:7]: But most of the time, users won't do that.

They will just:
```javascript
import Moai from '@moai-vue/core';

app.use(Moai);
```

Then use in template:

```vue
<template>
   <m-button></m-button>
</template>
```

Without any additional import for each view/component

Comment (thien-do) [reply lib/core/src/components/button/Button.vue:7]: yeah that sounds good!

Comment (iamquang95) [reply lib/core/src/components/button/Button.vue:7]: Oh #TIL: https://github.com/vuejs/vue-loader/issues/1234

> Features like hot module replacement, CSS injection, scoped CSS, SSR critical CSS collection all depend on being able to precisely locate the exported component - it will not work if the component can be exported as an arbitrarily named component. In addition, there are tools (ESLint plugin, for example) that are based on the assumption of the component being the default export.

> TL;DR: the component must be the default export and this is not going to change.
